### ISSUE TITLE
Add the "attestation simulator" performance metric

### ISSUE THREAD

**Issue by paulhauner** [sigp/lighthouse#4526]
## Description

I think it would be interesting if the BN would "pretend" to create an attestation every slot and then look at how that attestation would have performed.

In detail, I think we should call [`BeaconChain::produce_unaggregated_attestation`](https://github.com/sigp/lighthouse/blob/dfcb3363c757671eb19d5f8e519b4b94ac74677a/beacon_node/beacon_chain/src/beacon_chain.rs#L1557-L1561) 4s into each slot and store the resulting `attestation.data`. Then, in `ValidatorMonitor::process_valid_state` we could see if the `attestation.data` matches the head/target/source.

Some things to consider:

- If we use the `ValidatorMonitor` then it'll only run when the user has the validator monitor enabled. That seems reasonable to me.
- We probably want to only check the head/target/source once the `state.slot` is some slots beyond `attestation.data.slot` to defend against re-orgs. I'd suggest 16 slots as a trade-off between being safe enough and fast enough. I'm open to suggestion though!

### Attestation reward details

When I talk about "matching the head/target/source" I'm talking about whether the attestation meets the requirements to be given rewards. The [Altair specs](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/beacon-chain.md#participation-flag-indices) define the "Participation Flag Indices". If a validator has one of those flags set, then they receive rewards. You can see them getting set in [`get_attestation_participation_flag_indices`](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/beacon-chain.md#get_attestation_participation_flag_indices).

### Timer example

Here's an example of a task on the `BeaconChain` that runs periodically each slot:

https://github.com/sigp/lighthouse/blob/dfcb3363c757671eb19d5f8e519b4b94ac74677a/beacon_node/beacon_chain/src/proposer_prep_service.rs#L27-L72

**Comment by paulhauner** [comment]
I believe there is an EPF fellow working on this, FYI

**Comment by ameya-deshmukh** [comment]
@paulhauner yep, taking this up!

**Comment by v4lproik** [comment]
Hey @ameya-deshmukh Do you need any help on this or wanna collaborate?

**Comment by ameya-deshmukh** [comment]
Hey @v4lproik I'd prefer working on this alone :)

**Comment by v4lproik** [comment]
Hi @ameya-deshmukh,

First off, I couldn't see any of your updates on the EPF group so I really hope everything's alright with you :pray: and that I am not making a mistake as nickname / realname can be quite confusing.

I showed my interest in working on this topic a couple of months ago and as I do not really know the state of your work as I mentioned in the above, I decided to tackle this issue.

I hope that's okay with you and that if you'd like to collaborate, my offer still stands :)

I also hope that's okay with the LH team, I don't want to step on anyone's toes here.

**Comment by ameya-deshmukh** [comment]
Hey @v4lproik thanks so much for checking in and taking this up! Unfortunately I got a bit busy with college so I couldn't devote a lot of time to this project. I'd still like to collaborate in case you're interested :)

**Comment by v4lproik** [comment]
@ameya-deshmukh Maybe you can let me know what you have been working on and what you understand of the simulator so we can figure out what could be a good fit. I have basically read the eth2book / committee section and started implementing something the last week-end (cf. PR).

**Comment by v4lproik** [comment]
Hi LH team!

I am making progress with this issue but a few questions came to mind whilst coming up with a solution and I wanted to circle back with you and understand your view on the matter.

I am understanding that by producing an unaggregated attestation (UA) in a service running at each slot on the BN, we'll mainly be able to assess the performance of `BeaconChain::produce_unaggregated_attestation` as the complete flow would also involve modifying the VC to sign the UA such as:
1. VC calls the BN to get an UA
2. BN produces an UA and returns it to the VC
3. VC signes the UA
4. VC sends the signed attestations to the BN
5. BN validates and broadcasts the signed attestations

First off, I wanted to check with you that what's expected here is that we are not introducing any code in the VC?

Secondly, to my understanding, what I should do in order to check the head/target/source is pretty basic as shown in the PR as we are completely mocking the whole back and forth between the VC and the BN and that the attestation never gets included in the state (never included in a block). I decided to only throw an error message if the 3 flags are not set with a predefined inclusion_delay equals to min_inclusion_delay as specified in the [Altair specs](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/beacon-chain.md#participation-flag-indices). I leave it as is as there's no inclusion delay as we are not simulating any delay when producing an ua and that it is never inserted in a block as never signed by the VC, send back to... etc. Could you please tell me if I am on the right path here?

I was thinking of adding a prometheus metric counting each time when one of the 3 flags is not set. Let me know if that's something you'd find relevant or a log error is enough.

I was also thinking that it would be nice for the tests relying on the. `ValidatorMonitor` to be able to compare the Prometheus values set when `process_valid_state`. Is it something that you'd like to see implemented (scrap + exposing on http server) if it's scoped in another ticket? Would love to get on with it :)

Also and last question, are you planning on going beyond theis scope to implement > 32 attestation per epoch? (eg. making the VC submits more attestations to the BN etc..)

Thanks a lot for your time!

**Comment by paulhauner** [comment]
> First off, I wanted to check with you that what's expected here is that we are not introducing any code in the VC?

Yep, that's right. We don't need to add any code to the VC 👍 

> Could you please tell me if I am on the right path here?

I've given some feedback on the PR about this.

> I was also thinking that it would be nice for the tests relying on the. ValidatorMonitor to be able to compare the Prometheus values set when process_valid_state. Is it something that you'd like to see implemented (scrap + exposing on http server) if it's scoped in another ticket?

It could indeed be useful for testing! You might be able to scrape directly and bypass the HTTP server (see [`lighthouse_metrics::gather`](https://github.com/sigp/lighthouse/blob/441fc1691b69f9edc4bbdc6665f3efab16265c9b/common/lighthouse_metrics/src/lib.rs#L68-L71).

> Also and last question, are you planning on going beyond theis scope to implement > 32 attestation per epoch? 

I'm not sure what you mean here sorry.

**Comment by v4lproik** [comment]
Hi @paulhauner!

Sorry for coming back to you a little late, it's been slightly hectic lately with work and DevConnect.

However, I have kept on working on this and I think it can go through another round of feedbacks! Also, I have deployed the PR code version this morning on Goerli and came up with the graphs. I am opening a PR very soon on lighthouse-metrics if you are satisfied with the code. I also modified the graphs for "the missed blocks" according to the latest change. I will create and update the "missed blocks" PR in lighthouse-metrics repo at the same time then.    
  

>  I'm not sure what you mean here sorry.

Sorry for the confusion, I was asking if there was more work in the pipe around the BN as I would like to keep going :)
